lib/pharos/ssh/client.rb

```diff
@@ -157,58 +46,54 @@ def connect
         @session = Net::SSH.start(@host, @user, @opts)
       end
 
+      # @example
+      #   tempfile do |tmp|
+      #     exec!("less #{tmp}")
+      #   end
+      # @example
+      #   tmp = tempfile.new(content: "hello")
+      #   exec!("cat #{tmp}")
+      #   tmp.unlink
+      #
+      # @param prefix [String] tempfile filename prefix (default "pharos")
+      # @param content [String] initial file content, default blank
+      # @param file [String,IO] path to local file or a readable IO object
+      # @return [Pharos::SSH::Tempfile]
+      # @yield [Pharos::SSH::Tempfile]
+      def tempfile(prefix: "pharos", content: nil, file: nil, &block)
+        Tempfile.new(self, prefix: prefix, content: content, file: file, &block)
+      end
+
       # @param cmd [String] command to execute
-      # @return [Exec]
+      # @return [Pharos::Command::Result]
       def exec(cmd, **options)
         require_session!
-
-        logger.debug { "exec: #{cmd}" }
-
-        ex = Exec.new(cmd, **options)
-        ex.open(@session)
-        ex.wait
-        ex
+        RemoteCommand.new(self, cmd, **options).run
       end
 
       # @param cmd [String] command to execute
-      # @raise [ExecError]
+      # @raise [Pharos::SSH::RemoteCommand::ExecError]
       # @return [String] stdout
       def exec!(cmd, **options)
-        ex = exec(cmd, **options)
-
-        raise ExecError.new(cmd, ex.exit_status, ex.output) if ex.error?
-
-        ex.stdout
+        require_session!
+        RemoteCommand.new(self, cmd, **options).run!.stdout
       end
 
       # @param script [String] name of script
       # @param path [String] real path to file, defaults to script
-      # @raise [ExecError]
+      # @raise [Pharos::SSH::RemoteCommand::ExecError]
       # @return [String] stdout
       def exec_script!(name, env: {}, path: nil, **options)
         script = File.read(path || name)
-        cmd = ['sudo']
-
-        env.each_pair do |e, value|
-          cmd << "#{e}=#{Shellwords.escape(value)}"
-        end
-
-        cmd << 'sh'
-        cmd << '-x'
-
-        ex = exec(cmd.join(' '), stdin: script, debug_source: name, **options)
-
-        raise ExecError.new(name, ex.exit_status, ex.output) if ex.error?
-
-        ex.stdout
+        cmd = %w(sudo sh -x)
+        cmd.concat(env.map { |pair| pair.map(&:shellescape).join('=') })
```

Comment:
This line would be far less clever, and thus easier to read, if it was something like:

```ruby
cmd.concat(env.map{|key, value| "#{key.shellescape}=#{value.shellescape}" })
```

Using `map` for something known to be a fixed `key, value' pair feels wrong.

Reply:
Also the `String#shellescape`

```
    Configuring container runtime (docker) packages ...
undefined method `shellescape' for :DOCKER_PACKAGE:Symbol
/app/lib/pharos/ssh/client.rb:89:in `map'
/app/lib/pharos/ssh/client.rb:89:in `block in exec_script!'
/app/lib/pharos/ssh/client.rb:89:in `each'
/app/lib/pharos/ssh/client.rb:89:in `map'
/app/lib/pharos/ssh/client.rb:89:in `exec_script!'
/app/lib/pharos/phases/base.rb:17:in `exec_script'
/app/lib/pharos/phases/configure_host.rb:38:in `call'
/app/lib/pharos/up_command.rb:159:in `handle_masters'
/app/lib/pharos/up_command.rb:90:in `block in configure'
/app/lib/pharos/up_command.rb:89:in `chdir'
/app/lib/pharos/up_command.rb:89:in `configure'
/app/lib/pharos/up_command.rb:41:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

Reply:
Needs `key.to_s.shellescape`